contracts/src/scripts/integration-test.ts

```diff
@@ -0,0 +1,176 @@
+});
+
+const fee = Number(config.fee) * 1e9; // in nanomina (1 billion = 1.0 mina)
+Mina.setActiveInstance(Network);
+let tx;
+let feepayerAddress = feepayerKey.toPublicKey();
+console.log(feepayerAddress.toBase58());
+let zkAppAddress = zkAppKey.toPublicKey();
+let name_service_contract = new NameService(zkAppAddress);
+
+console.time('compile program');
+await offchainState.compile();
+offchainState.setContractInstance(name_service_contract);
+console.timeEnd('compile program');
+console.time('compile contract');
+await NameService.compile();
+console.timeEnd('compile contract');
+
+console.time('deploy');
+tx = await Mina.transaction({ sender: feepayerAddress, fee: fee }, async () => {
+  AccountUpdate.fundNewAccount(feepayerAddress);
+  await name_service_contract.deploy();
+})
+  .prove()
+  .sign([feepayerKey, zkAppKey])
+  .send()
+  .wait();
+console.timeEnd('deploy');
+
+console.time('set premimum rate');
+tx = await Mina.transaction({ sender: feepayerAddress, fee: fee }, async () => {
+  await name_service_contract.set_premium(UInt64.from(100));
+})
+  .sign([feepayerKey])
+  .prove()
+  .send()
+  .wait();
+console.timeEnd('set premimum rate');
+
+console.time('settlement proof 1');
+let proof = await offchainState.createSettlementProof();
+console.timeEnd('settlement proof 1');
+
+console.time('settle 1');
+tx = await Mina.transaction({ sender: feepayerAddress, fee: fee }, async () =>
+  name_service_contract.settle(proof)
+)
+  .sign([feepayerKey])
+  .prove()
+  .send()
+  .wait();
+console.timeEnd('settle 1');
+
+console.time('get premimum rate');
+let res;
+tx = await Mina.transaction({ sender: feepayerAddress, fee: fee }, async () => {
+  res = await name_service_contract.premium_rate();
+})
+  .sign([feepayerKey])
+  .prove()
+  .send()
+  .wait();
+console.log(res!.toString());
+console.timeEnd('get premimum rate');
+
+for (let i = 0; i < cycleNumber; i++) {
+  for (let j = 0; j < 3; j++) {
+    let name = Math.random().toString(36).substring(2, 12).concat('.mina');
+    let new_record = new NameRecord({
+      mina_address: PrivateKey.randomKeypair().publicKey,
+      avatar: Field.random(),
+      url: Field.random(),
+    });
+    names.push(name);
+    nameMap.set(name, new_record);
+
+    console.time('register a name');
+    tx = await Mina.transaction(
+      { sender: feepayerAddress, fee: fee + 100 },
+      async () => {
+        await name_service_contract.register_name(
+          Name.fromString(name),
+          new_record
+        );
+      }
+    )
+      .sign([feepayerKey])
+      .prove()
+      .send()
+      .wait();
+    console.timeEnd('register a name');
+  }
+
+  wait(8); // wait for settlement
+
+  console.time('get a randomName');
+  let randomName = names[Math.floor(Math.random() * names.length)];
+  let record = nameMap.get(randomName);
+  tx = await Mina.transaction(
+    { sender: feepayerAddress, fee: fee },
+    async () => {
+      let res = await name_service_contract.resolve_name(
```

Comment:
My script failed here since the name was not owned.  I think the problem is `wait(8); // wait for settlement`.  This only waits for 8 seconds, which doesn't seem like long enough.

I think the wait function should be upgraded to watch the transaction we created, and only proceed when the transaction is applied.  No matter how long we wait, the network may not yet have applied the transaction.

Reply:
yeah, wait function is incorrect. I probably missed that because the random case would have been in the set of settled ones. 

Watching the settlement transaction would be the best but I don't think we can do that in o1js.